src/overlay-manager.coffee

```diff
@@ -1,38 +1,101 @@
 
   render: (state) ->
-    for decorationId, {pixelPosition, item} of state.content.overlays
-      @renderOverlay(state, decorationId, item, pixelPosition)
+    editorDimensionsHaveChanged = !@editorDimensionsAreEqual(state)
 
-    for id, overlayNode of @overlayNodesById
+    for decorationId, overlay of state.content.overlays
+      overlayHasChanged = not @overlayStateIsEqual(decorationId, overlay)
+      if editorDimensionsHaveChanged or overlayHasChanged
+        @renderOverlay(state, decorationId, overlay)
+        @cacheOverlayState(decorationId, overlay)
+
+    for id, {overlayNode} of @overlaysById
       unless state.content.overlays.hasOwnProperty(id)
-        delete @overlayNodesById[id]
+        delete @overlaysById[id]
         overlayNode.remove()
 
-    return
+    @cacheEditorDimensions(state)
+
+  overlayStateIsEqual: (decorationId, overlay) ->
+    return false unless @overlaysById[decorationId]?
+    @overlaysById[decorationId].itemWidth is overlay.itemWidth and
+      @overlaysById[decorationId].itemHeight is overlay.itemHeight and
+      @overlaysById[decorationId].contentMargin is overlay.contentMargin and
+      @overlaysById[decorationId].pixelPosition?.top is overlay.pixelPosition?.top and
+      @overlaysById[decorationId].pixelPosition?.left is overlay.pixelPosition?.left
+
+  cacheOverlayState: (decorationId, overlay) ->
+    return unless @overlaysById[decorationId]?
+    @overlaysById[decorationId].itemWidth = overlay.itemWidth
+    @overlaysById[decorationId].itemHeight = overlay.itemHeight
+    @overlaysById[decorationId].contentMargin = overlay.contentMargin
+    @overlaysById[decorationId].pixelPosition = overlay.pixelPosition
+
+  cacheEditorDimensions: (state) ->
+    @cachedEditorDimensions =
+      lineHeight: @presenter.lineHeight
+      contentFrameWidth: @presenter.contentFrameWidth
+      editorTop: @presenter.boundingClientRect?.top
+      editorLeft: @presenter.boundingClientRect?.left
+      editorWidth: @presenter.boundingClientRect?.width
+      windowWidth: @presenter.windowWidth
+      windowHeight: @presenter.windowHeight
+      scrollTop: state.content.scrollTop
+      scrollLeft: state.content.scrollLeft
+
+  editorDimensionsAreEqual: (state) ->
+    return false unless @cachedEditorDimensions?
+    @cachedEditorDimensions.lineHeight is @presenter.lineHeight and
+      @cachedEditorDimensions.contentFrameWidth is @presenter.contentFrameWidth and
+      @cachedEditorDimensions.editorTop is @presenter.boundingClientRect?.top and
+      @cachedEditorDimensions.editorLeft is @presenter.boundingClientRect?.left and
+      @cachedEditorDimensions.editorWidth is @presenter.boundingClientRect?.width and
+      @cachedEditorDimensions.windowWidth is @presenter.windowWidth and
+      @cachedEditorDimensions.windowHeight is @presenter.windowHeight and
+      @cachedEditorDimensions.scrollTop is state.content.scrollTop and
+      @cachedEditorDimensions.scrollLeft is state.content.scrollLeft
```

Comment:
We _could_ ignore all this caching, and render them all the time...

If keeping the cache, maybe it goes in the presenter?